File: programmers/level2/프렌즈4블록.java
import java.util.*;

class Solution {
	void print(char[][] arr) {
		for(int i = 0; i < arr.length; ++i) {
			System.out.println(Arrays.toString(arr[i]));
		}
		System.out.println();
	}
	public int solution(int m, int n, String[] board) {
		int answer = 0;
		char[][] arr = new char[m][n];
		for (int i = 0; i < m; ++i) {
			arr[i] = board[i].toCharArray();
		}
		while (true) {
			boolean[][] check = new boolean[m][n];
			int count = 0;
			for (int i = 0; i < m - 1; ++i) {
				for (int j = 0; j < n - 1; ++j) {
					if (arr[i][j] != ' ' && arr[i][j] == arr[i][j + 1] && arr[i][j] == arr[i + 1][j]
							&& arr[i + 1][j] == arr[i + 1][j + 1]) {
						check[i][j] = check[i][j + 1] = check[i + 1][j] = check[i + 1][j + 1] = true;
						count++;
					}
				}
			}
			if (count == 0) {
				break;
			}
			for (int j = 0; j < n; ++j) {
				Queue<Character> q = new LinkedList<>();
				for (int i = m - 1; i >= 0; --i) {
					if (!check[i][j]) {
						q.add(arr[i][j]);
					}
				}
				int idx = m - 1;
				while (!q.isEmpty()) {
					arr[idx--][j] = q.poll();
				}
				while (idx >= 0) {
					arr[idx--][j] = ' ';
				}
			}
		}
		for(int i = 0; i < m; ++i) {
			for(int j = 0; j < n; ++j) {
				if(arr[i][j] == ' ') {
					answer++;
				}
			}
		}
		return answer;
	}
}

public class 프렌즈4블록 {
	public static void main(String[] args) {
		int m = 4;
		int n = 5;
		String[] board = { "CCBDE", "AAADE", "AAABF", "CCBBF" };
		Solution sol = new Solution();
		System.out.println(sol.solution(m, n, board));
	}
}

File: programmers/python/level2/프렌즈4블록.py
def solution(m, n, board):
    board = [[c for c in row] for row in board]
    answer = 0
    while True:
        result = check_remove(board)
        count, remove_check = result[0], result[1]
        if count == 0:
            break
        answer += count
        down(board, remove_check)
    return answer

def check_remove(board):
    remove_check = [[False for j in range(len(board[0]))] for i in range(len(board))]
    count = 0
    for i in range(len(board)-1):
        for j in range(len(board[0])-1):
            if board[i][j] != 0 and board[i][j] == board[i][j+1] == board[i+1][j] == board[i+1][j+1]:
                remove_check[i][j] = remove_check[i][j+1] = remove_check[i+1][j] = remove_check[i+1][j+1] = True

    for i in range(len(remove_check)):
        for j in range(len(remove_check[0])):
            if remove_check[i][j]:
                count += 1
    return [count, remove_check]

def down(board, remove_check):
    for j in range(len(board[0])):
        temp = []
        for i in range(len(board)):
            if remove_check[i][j] == False:
                temp.append(board[i][j])
        for i in range(len(board)-len(temp)):
            board[i][j] = 0
        for i in range(len(board)-1, len(board)-1-len(temp), -1):
            board[i][j] = temp.pop()

print(solution(4,	5	,["CCBDE", "AAADE", "AAABF", "CCBBF"]))